homeassistant/components/seventeentrack/sensor.py

```diff
@@ -210,51 +202,66 @@ def state(self):
     @property
     def unique_id(self):
         """Return a unique, HASS-friendly identifier for this entity."""
-        return ENTITY_ID_TEMPLATE.format(
+        return UNIQUE_ID_TEMPLATE.format(
             self._data.account_id, self._tracking_number)
 
     async def async_update(self):
         """Update the sensor."""
         await self._data.async_update()
 
-        if not self._data.packages:
+        if not self.available:
+            await self._remove()
             return
 
-        try:
-            package = next((
-                p for p in self._data.packages
-                if p.tracking_number == self._tracking_number))
-        except StopIteration:
-            # If the package no longer exists in the data, log a message and
-            # delete this entity:
-            _LOGGER.info(
-                'Deleting entity for stale package: %s', self._tracking_number)
-            reg = await self.hass.helpers.entity_registry.async_get_registry()
-            self.hass.async_create_task(reg.async_remove(self.entity_id))
-            self.hass.async_create_task(self.async_remove())
-            return
+        package = self._data.packages.get(self._tracking_number, None)
 
-        # If the user has elected to not see delivered packages and one gets
+        # If the user has selected to not see delivered packages and one gets
         # delivered, post a notification:
         if package.status == VALUE_DELIVERED and not self._data.show_delivered:
-            _LOGGER.info('Package delivered: %s', self._tracking_number)
-            self.hass.components.persistent_notification.create(
-                'Package Delivered: {0}<br />'
-                'Visit 17.track for more infomation: {1}'
-                ''.format(
-                    self._tracking_number,
-                    NOTIFICATION_DELIVERED_URL_SCAFFOLD.format(
-                        self._tracking_number)),
-                title=NOTIFICATION_DELIVERED_TITLE,
-                notification_id=NOTIFICATION_DELIVERED_ID_SCAFFOLD.format(
-                    self._tracking_number))
+            self._notify_delivered()
+            await self._remove()
             return
 
         self._attrs.update({
             ATTR_INFO_TEXT: package.info_text,
             ATTR_LOCATION: package.location,
         })
         self._state = package.status
+        self._friendly_name = package.friendly_name
+
+    def _async_write_ha_state(self):
+        """Avoid writing state, so the entity is really removed."""
+        if not self._removed:
+            return super()._async_write_ha_state()
```

Comment:
We shouldn't be calling a private method on the `Entity` class.

Reply:
Yes, we shouldn't overwrite this private entity method either.

Reply:
Hm there is something I don't get:
First intention was to prevent HA to write the state if entity was removed. Why ? Because if the state was written, then entity was still there and flagged is unavailable instead of being completely deleted.

I removed the override and replaced `await self._remove()` by `self.hass.async_create_task(self._remove())` and it seems to work as intended (entity is completely removed).

Could you confirm I'm right with the logic:
with `await self._remove()`:
- entity update occurs
- entity is removed during the update
- after the update the new state is written -> entity is added again

with `self.hass.async_create_task(self._remove())`
- entity update occurs
- after the update the new state is written
- entity is removed (`self._remove()` is called)

Am I right ?

Reply:
I'm not the expert on how the state machine works, but that looks correct to me. 👍

Reply:
Yes, that's correct. We need to schedule the removal after the state update is completed, to not have the state update add the entity back after removing it.